Refresh.GameServer/Database/GameDatabaseContext.Relations.cs

```diff
@@ -192,6 +192,81 @@ public void ClearQueue(GameUser user)
 
     #region Rating and Reviewing
 
+    public void RateReview(GameReview review, RatingType rt, GameUser user)
+    {
+        // if rating type is neutral, remove previous review rating by user 
+        if (rt == RatingType.Neutral && this.ReviewRatingExistsByUser(user, review))
+        {
+            RateReviewRelation rel =
+                this.RateReviewRelations.First(relation => relation.Review == review && relation.User == user);
+            this.Write(() => this.RateReviewRelations.Remove(rel));
+
+            return;
+        }
+        // swap rating
+        if (this.ReviewRatingExistsByUser(user, review))
+        {
+            RateReviewRelation rel = this.RateReviewRelations.First(relation => relation.Review == review && relation.User == user);
+            
+            this.Write(() => rel.RatingType = rt);
+
+            return;
+        }
+        
+        RateReviewRelation reviewRelation = new()
+        {
+            RatingType = rt,
+            Review = review,
+            User = user,
+        };
+
+        this.Write(() =>
+        {
+            this.RateReviewRelations.Add(reviewRelation);
+        });
+    }
+    
+    public ReviewRating GetRatingForReview(GameReview review)
+    {
+        IQueryable<RateReviewRelation> relations = this.RateReviewRelations.Where((relation => relation.Review == review));
+        ReviewRating rating = new();
+
+        foreach (RateReviewRelation rel in relations)
+        {
+            if (rel.RatingType == RatingType.Yay)
+            {
+                rating.PostiveRating++;
+            }
+            else
+            {
+                rating.NegativeRating++;
+            }
+        }
+
+        return rating;
+    }
+
+    public GameReview GetReviewByUsernameForLevelId(string userName, int levelId)
+    {
+        GameLevel level = this.GameLevels.First(gameLevel => gameLevel.LevelId == levelId);
+        GameUser user = this.GameUsers.First(gameUser => gameUser.Username == userName);
+        
+        GameReview review = this.GameReviews.First((gameReview => gameReview.Publisher == user && gameReview.Level == level));
+        
+        return review;
+    }
```

Comment:
In cases where you haven't verified if a review actually exists, you should use `FirstOrDefault` and handle nullable cases like this:
```csharp
    public GameReview? GetReviewByUsernameForLevelId(string userName, int levelId)
    {
        GameLevel? level = this.GameLevels.FirstOrDefault(gameLevel => gameLevel.LevelId == levelId);
        if(level == null) return null;
        GameUser? user = this.GameUsers.FirstOrDefault(gameUser => gameUser.Username == userName);
        if(user == null) return null;
        
        GameReview? review = this.GameReviews.FirstOrDefault(gameReview => gameReview.Publisher == user && gameReview.Level == level);
        
        return review;
    }
```

But I wouldn't do that in this case; I would take `GameUser` and `GameLevel` in directly; typically endpoints handle missing elements. But definitely handle the case where the end `GameReview` doesn't exist.